{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "<script>\n",
    "  jQuery(document).ready(function($) {\n",
    "\n",
    "  $(window).load(function(){\n",
    "    $('#preloader').fadeOut('slow',function(){$(this).remove();});\n",
    "  });\n",
    "\n",
    "  });\n",
    "</script>\n",
    "\n",
    "<style type=\"text/css\">\n",
    "  div#preloader { position: fixed;\n",
    "      left: 0;\n",
    "      top: 0;\n",
    "      z-index: 999;\n",
    "      width: 100%;\n",
    "      height: 100%;\n",
    "      overflow: visible;\n",
    "      background: #fff url('http://preloaders.net/preloaders/720/Moving%20line.gif') no-repeat center center;\n",
    "  }\n",
    "\n",
    "</style>\n",
    "\n",
    "<div id=\"preloader\"></div>"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "<script>\n",
    "  function code_toggle() {\n",
    "    if (code_shown){\n",
    "      $('div.input').hide('500');\n",
    "      $('#toggleButton').val('Show Code')\n",
    "    } else {\n",
    "      $('div.input').show('500');\n",
    "      $('#toggleButton').val('Hide Code')\n",
    "    }\n",
    "    code_shown = !code_shown\n",
    "  }\n",
    "\n",
    "  $( document ).ready(function(){\n",
    "    code_shown=false;\n",
    "    $('div.input').hide()\n",
    "  });\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" id=\"toggleButton\" value=\"Show Code\"></form>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load rpy2 magic\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "# to switch off warning messages\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# make default cell width 85% of available screen\n",
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:85% !important; }</style>\"))\n",
    "\n",
    "# load R libraries & functions\n",
    "%R options(warn=-1)\n",
    "%R library(RColorBrewer)\n",
    "%R library(ComplexHeatmap)\n",
    "%R library(circlize)\n",
    "%R library(dendextend)\n",
    "%R library(Rtsne)\n",
    "%R library(ggplot2)\n",
    "%R library(gplots)\n",
    "%R library(gridExtra)\n",
    "%R library(DESeq2)\n",
    "%R library(dendextend)\n",
    "%R library(ggrepel)\n",
    "    \n",
    "# load python modules\n",
    "import glob\n",
    "import re\n",
    "import sys\n",
    "import os\n",
    "import rpy2.robjects as robjects\n",
    "import CGAT.Database as DB\n",
    "import sqlite3\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "db = \"./csvdb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# R functions\n",
    "theme_notebook <- function(base_size=18, base_family=\"helvetica\") {\n",
    "                  (theme_set(theme_minimal(base_size=18))\n",
    "                  + theme(plot.title = element_text(face=\"bold\", size=20, hjust=0.5),\n",
    "                             text = element_text(),\n",
    "                             axis.title = element_text(face=\"bold\",size = rel(1)),\n",
    "                             axis.title.y = element_text(angle=90,vjust=2, size=20),\n",
    "                             axis.title.x = element_text(vjust=-0.2, size=20),\n",
    "                             axis.text = element_text(size=20),\n",
    "                             axis.line = element_line(colour=\"black\"),\n",
    "                             axis.ticks = element_line(),\n",
    "                             legend.key = element_rect(colour = NA),\n",
    "                             legend.key.size= unit(0.5, \"cm\"),\n",
    "                             legend.margin = unit(0.5, \"cm\"),\n",
    "                             legend.text = element_text(size=14),\n",
    "                             legend.title = element_text(size=16),\n",
    "                             ))\n",
    "}\n",
    "\n",
    "ggplot_prcomp <- function(prcomp_object,\n",
    "                          plots=list(\"A\"=c(\"PC1\",\"PC2\"), \"B\"=c(\"PC3\",\"PC4\"), \"C\"=c(\"PC5\",\"PC6\")),\n",
    "                          sample_information=\"none\",\n",
    "                          color=\"c()\",\n",
    "                          shape=\"c()\",\n",
    "                          label=\"none\",\n",
    "                          size=3,\n",
    "                          nudge_scale_factor=40){\n",
    "    require(gridExtra)\n",
    "    pca = prcomp_object\n",
    "\n",
    "    # sample_information should have the same rownames as pca$x\n",
    "    pvs <- summary(pca)$importance[\"Proportion of Variance\",]\n",
    "\n",
    "    names = paste(names(pvs),\" (\",round(pvs,2),\")\",sep=\"\")\n",
    "\n",
    "    #scree plot\n",
    "    fs <- data.frame(x=c(1:length(names(pvs))), y=as.vector(pvs))\n",
    "\n",
    "    pcdf <- as.data.frame(pca$x)\n",
    "\n",
    "    pcdf <- merge(pcdf, sample_information, by=0, all=T)\n",
    "\n",
    "    gps = list()\n",
    "\n",
    "    scree <- ggplot(fs, aes(x,y)) + \n",
    "                geom_point(size=4) + \n",
    "                xlab(\"principal component\") + \n",
    "                ylab(\"proportion of variance\") + \n",
    "                ggtitle(\"scree plot\")\n",
    "\n",
    "    c_lab <- function(props, C){\n",
    "        return(paste(C, \" (\", props[[C]]*100,\"%)\",sep=\"\"))\n",
    "    }\n",
    "\n",
    "    for(plot in names(plots)){\n",
    "\n",
    "        comps <- plots[[plot]]\n",
    "\n",
    "        PCX <- comps[1]\n",
    "        PCY <- comps[2]\n",
    "\n",
    "        nudge_x <- diff(range(pcdf[[PCX]]))/nudge_scale_factor\n",
    "        nudge_y <- diff(range(pcdf[[PCY]]))/nudge_scale_factor\n",
    "\n",
    "        gp <- ggplot(pcdf, aes_string(PCX, PCY, color=color, shape=shape))\n",
    "\n",
    "        if(label!=\"none\"){\n",
    "            gp <- gp + geom_text(aes_string(label=label), nudge_x=nudge_x, nudge_y=nudge_y, color=\"black\")\n",
    "        }\n",
    "\n",
    "        gp <- gp + geom_point(size=size)\n",
    "        gp <- gp + xlab(c_lab(pvs,PCX)) + ylab(c_lab(pvs,PCY))\n",
    "\n",
    "        gps[[plot]] <- gp\n",
    "\n",
    "    }\n",
    "\n",
    "    gps[[\"scree\"]] <- scree\n",
    "\n",
    "    return(gps)\n",
    "}\n",
    "\n",
    "plot_heatmap <- function(matrix, clust_method=\"ward.D2\", dist_method=\"manhattan\", low_opt=0.1, \n",
    "             med_opt=0.5, high_opt=0.9, labels=\"random\", no_labels=50){\n",
    "\n",
    "    # get vector of labels\n",
    "    if (labels == \"random\"){\n",
    "        subset <- matrix[sample(no_labels, replace=FALSE), ]\n",
    "        subset <- as.integer(rownames(subset))\n",
    "        labels <- matrix$gene_name[subset]\n",
    "    } else{\n",
    "        # modify this to optionally label top DE genes\n",
    "        # or most accessible peaks\n",
    "        print(\"Specify what to label, options=random\")\n",
    "    }\n",
    "\n",
    "    rnames <- matrix$gene_name\n",
    "    cnames <- sample_info$category\n",
    "    matrix$gene_name <- NULL\n",
    "\n",
    "    matrix <- data.matrix(matrix)\n",
    "    rownames(matrix) <- rnames\n",
    "    colnames(matrix) <- cnames\n",
    "\n",
    "    # Calculate distance\n",
    "    if(dist_method == \"minkowski\"){\n",
    "        p=1.44 # If method is \"minkowski\" set p as 1.44\n",
    "        dist_genes <- dist(matrix, method=dist_method, p)\n",
    "        }\n",
    "    else{\n",
    "        dist_genes <- dist(matrix, method=dist_method)\n",
    "        }\n",
    "\n",
    "    # Clustering\n",
    "    clust_genes <- hclust(dist_genes, method=clust_method)\n",
    "    den_genes <- as.dendrogram(clust_genes)\n",
    "\n",
    "    if(dist_method == \"minkowski\"){\n",
    "         p=1.44\n",
    "         dist_samples <- dist(t(matrix), method=dist_method, p)\n",
    "        }\n",
    "    else{\n",
    "         dist_samples <- dist(t(matrix), method=dist_method)\n",
    "        }\n",
    "\n",
    "    clust_samples <- hclust(dist_samples, method=clust_method)\n",
    "    den_samples <- as.dendrogram(clust_samples)\n",
    "\n",
    "    # colour breaks\n",
    "    low <- quantile(matrix, low_opt)\n",
    "    med <- quantile(matrix, med_opt)\n",
    "    high <- quantile(matrix, high_opt)\n",
    "\n",
    "    ht = Heatmap(matrix, \n",
    "            col = colorRamp2(c(low, med, high), c(\"royalblue3\", \"white\", \"red3\"), space=\"LAB\"),\n",
    "            name=\"Accessibility\",\n",
    "            cluster_rows = den_genes,\n",
    "            cluster_columns = den_samples,\n",
    "            column_dend_height = unit(2, \"cm\"),\n",
    "            row_dend_width = unit(3, \"cm\"),\n",
    "            show_row_names = FALSE, \n",
    "            column_names_gp=gpar(fontsize=18),\n",
    "            heatmap_legend_param=list(legend_direction=\"horizontal\", \n",
    "                                      color_bar = \"continuous\",\n",
    "                                      legend_width = unit(5, \"cm\"), \n",
    "                                      title_position = \"lefttop\"),\n",
    "            show_column_dend = TRUE) + \n",
    "            rowAnnotation(link=row_anno_link(at=subset, labels=labels), width=unit(1, \"cm\") + \n",
    "                          max_text_width(labels))\n",
    "\n",
    "    draw(ht, heatmap_legend_side = \"bottom\")\n",
    "}\n",
    "\n",
    "# Set ggplot theme\n",
    "theme_set(theme_notebook(base_size=18))\n",
    "Palette <- c(\"#E69F00\", \"#0072B2\", \"#D55E00\", \"#009E73\", \"#56B4E9\",  \"#999999\", \"#F0E442\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **ATAC-seq - Differential Accessibility**\n",
    "***\n",
    "* Analysis based upon a merged peakset of all detected ATAC peaks (either from all-fragments or size-filtered)\n",
    "    * specified in pipeline.ini\n",
    "* Reads were counted over merged set of intervals using bedtools multicov \n",
    "* Raw counts are analysed for differential accessibility with DESeq2\n",
    "* Significant differentially accessible peaks used for data exploration "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get peak/readcounts to use from pipeline.ini\n",
    "with open(\"pipeline.ini\", \"r\") as o:\n",
    "    for line in o:\n",
    "        match = re.findall(\"^peaks=.*\", line)\n",
    "        if len(match) > 0:  \n",
    "            filt = ''.join([x.split(\"=\")[-1] for x in match])\n",
    "\n",
    "# import counts & upper quantile normalise\n",
    "def get_counts(filt=filt, db=db):\n",
    "    # get counts, filter on fragment size\n",
    "\n",
    "    if filt == \"size_filt\":\n",
    "        peaks = \"<150bp\"\n",
    "    if filt == \"all\":\n",
    "        peaks = \"all_fragments\"\n",
    "        \n",
    "    statement = '''select sample_id, peak_id, total from all_norm_counts\n",
    "                where size_filt == \"%(peaks)s\" ''' % locals()\n",
    "        \n",
    "    df = DB.fetch_DataFrame(statement, db)\n",
    "\n",
    "    df = df.pivot(\"sample_id\", \"peak_id\", \"total\").transpose()\n",
    "    df.index.name = None\n",
    "\n",
    "    return df\n",
    "\n",
    "counts = get_counts()\n",
    "# counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **DESeq2**\n",
    "### Sample information table:\n",
    "* \"condition\" column used for DESeq2 design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "sample_info = DB.fetch_DataFrame('''select * from sample_info''', db)\n",
    "sample_info.index = sample_info[\"sample_id\"]\n",
    "sample_info.index.name = None\n",
    "sample_info.head(len(sample_info))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variance stabilizing transformation (VST) \n",
    "* Count data are normalised by division by the size factors\n",
    "* Size factors are based on the fitted dispersion-mean relations\n",
    "* The transformation also normalises with respect to library size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " %%R -i counts,sample_info -w 600\n",
    "\n",
    "require(vsn)\n",
    "\n",
    "# replicates as characters\n",
    "sample_info$replicate <- as.character(sample_info$replicate)\n",
    "\n",
    "# build a DESeqDataSet from the count matrix\n",
    "# sample_info must have same rownames as colnames in count matrix\n",
    "dds <- DESeqDataSetFromMatrix(countData=counts, \n",
    "                              colData= sample_info, \n",
    "                              design = ~ condition )\n",
    "\n",
    "dds <- estimateSizeFactors(dds)\n",
    "\n",
    "# apply VST\n",
    "vstMat <- assay(varianceStabilizingTransformation(dds, blind=TRUE, fitType=\"local\"))\n",
    "colnames(vstMat) <- rownames(sample_info)\n",
    "notAllZero <- (rowMeans(vstMat)> min(vstMat))\n",
    "\n",
    "meanSdPlot(vstMat[notAllZero,], ylab=\"sd, vst\",plot=T)\n",
    "\n",
    "# run DESeq2\n",
    "dds <- DESeq(dds,  fitType=\"local\")\n",
    "# res <- results(dds, cooksCutoff=FALSE)\n",
    "res <- results(dds)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check dispersion estimate fit \n",
    "* default fit type = \"local\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "plotDispEsts(dds) # check dispersion estimate fit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DESeq2 Differential Accessibility Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%R summary(res)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DESeq2 QC Metrics\n",
    "\n",
    "### Cook's Distance\n",
    "* Diagnostic test for outliers, calculated for each gene and sample \n",
    "* Cooks distance is a measure of how much a single sample is influencing fitted coefficients for a gene, large values indicate an outlier count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -w 600 -h 400\n",
    "\n",
    "plot_cooks <- function(dds, sample_info){\n",
    "    \n",
    "    require(reshape)\n",
    "    require(reshape2)\n",
    "    \n",
    "    # DESeq function calculates Cooks distance (diagnostic test for outliers) for every gene and sample. \n",
    "    \n",
    "    # Cooks distance is a measure of how much a single sample is influencing fitted coefficients for a gene, \n",
    "    # large values indicate an outlier count\n",
    "\n",
    "    cooks <- melt(as.data.frame(log10(assays(dds)[[\"cooks\"]])))\n",
    "    cooks <- merge(cooks, sample_info, by.x=\"variable\", by.y=0, all=T)\n",
    "    \n",
    "    p <- ggplot(cooks, aes(y=value, x=variable, colour=condition)) +  \n",
    "        geom_boxplot()  + \n",
    "        theme(axis.text.x=element_blank()) + \n",
    "        labs(y=\"Cooks Distance (log10)\", x=\"\") +\n",
    "        scale_colour_manual(values=Palette)\n",
    "    \n",
    "    return(p)\n",
    "}\n",
    "\n",
    "a <- plot_cooks(dds, sample_info)\n",
    "a"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count Normalisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -w 800 -h 800 -o filt_norm_counts\n",
    "\n",
    "plot_countNorm <- function(dds, sample_info){\n",
    "    \n",
    "    # raw counts\n",
    "    raw <- melt(t(counts(dds)))\n",
    "    raw <- merge(raw, sample_info, by.x=\"X1\", by.y=0, all=T)\n",
    "\n",
    "    p1 <- ggplot(raw, aes(y=log2(value), x=X1, colour=condition)) + \n",
    "            geom_boxplot() + \n",
    "            theme(axis.text.x=element_blank()) +\n",
    "            labs(title=\"Raw counts\", x=\"\", y=\"\") + \n",
    "            scale_colour_manual(values=Palette)\n",
    "\n",
    "    # norm counts = counts(dds) / sizeFactors(dds)\n",
    "    norm <- melt(t(counts(dds, normalized=T)))\n",
    "    norm <- merge(norm, sample_info, by.x=\"X1\", by.y=0, all=T)\n",
    "\n",
    "    p2 <- ggplot(norm, aes(y=log2(value), x=X1, colour=condition)) + \n",
    "            geom_boxplot() + \n",
    "            theme(axis.text.x=element_blank()) +\n",
    "            labs(title=\"Normalised counts\", y=\"\", x=\"\") + \n",
    "            scale_colour_manual(values=Palette)\n",
    "\n",
    "    # VST counts\n",
    "    colnames(vstMat) <- dds@colData@rownames\n",
    "    data <-melt(vstMat[notAllZero,])\n",
    "    data <- merge(data, sample_info, by.x=\"X2\", by.y=0, all=T)\n",
    "\n",
    "    p3 <- ggplot(data, aes(y=value, x=X2, colour=condition)) + \n",
    "            geom_boxplot()  + \n",
    "            theme(axis.text.x=element_blank()) +\n",
    "            labs(title=\"VST counts\", y=\"\", x=\"\") + \n",
    "            scale_colour_manual(values=Palette) \n",
    "\n",
    "    # normalised counts with excluded counts (padj == NA) removed\n",
    "    # counts excluded by: Cooks cutoff & independent filtering\n",
    "    filtered_genes <- subset(res, res@listData$padj != \"NA\")@\"rownames\" # get sig DE gene names\n",
    "    filt_norm_counts <- subset(melt(t(counts(dds,normalized=T))), X2 %in% filtered_genes)\n",
    "\n",
    "    filt_norm_counts_info <- merge(filt_norm_counts, sample_info, by.x=\"X1\", by.y=0, all=T)\n",
    "\n",
    "    p4 <- ggplot(filt_norm_counts_info, aes(y=log2(value), x=X1, colour=condition)) + \n",
    "            geom_boxplot() + \n",
    "            theme(axis.text.x=element_blank()) +\n",
    "            labs(title=\"Filtered normalised counts\", x=\"\", y=\"\") + \n",
    "            scale_colour_manual(values=Palette) +\n",
    "            theme(legend.direction=\"horizontal\")\n",
    "    \n",
    "    get_legend <- function(a.gplot){ \n",
    "      tmp <- ggplot_gtable(ggplot_build(a.gplot)) \n",
    "      leg <- which(sapply(tmp$grobs, function(x) x$name) == \"guide-box\") \n",
    "      legend <- tmp$grobs[[leg]] \n",
    "      return(legend)}\n",
    "                      \n",
    "    key <- get_legend(p4)\n",
    "\n",
    "    plots = list(p1, p2, p3, p4, key, filt_norm_counts)\n",
    "\n",
    "    return(plots)\n",
    "}\n",
    "\n",
    "plots <- plot_countNorm(dds, sample_info)\n",
    "\n",
    "a <- plots[[1]]\n",
    "b <- plots[[2]]\n",
    "c <- plots[[3]]\n",
    "d <- plots[[4]]\n",
    "e <- plots[[5]]\n",
    "\n",
    "filt_norm_counts <- plots[[6]]\n",
    "\n",
    "grid.arrange(a + theme(legend.position=\"none\"), \n",
    "             b + theme(legend.position=\"none\"), \n",
    "             d + theme(legend.position=\"none\"), \n",
    "             c + theme(legend.position=\"none\"), \n",
    "             bottom=e, ncol=2, nrow=2,\n",
    "             left=textGrob(\"log2(counts)\", gp=gpar(fontsize=21, fontfamily=\"Helvetica\"), rot=90))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# rearrange data -> by gene\n",
    "filt_norm_counts = filt_norm_counts.pivot(\"X2\", \"X1\", \"value\")\n",
    "filt_norm_counts.index.name = None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Relative Log Expression (RLE)\n",
    "* Normalises samples for size factors\n",
    "* Briefly, the gene expression in each sample is normalised to the median expression of that gene across all samples & counts are log2 transformed\n",
    "* RLE should be centered over 0, dispersion from this indicates poor quality normalisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%R -w 800 -h 800 -i filt_norm_counts\n",
    "\n",
    "plot_countRLE <- function(dds, sample_info, filt_norm_counts){\n",
    "\n",
    "    # raw counts\n",
    "    rawc <- apply(counts(dds),1,median)\n",
    "    rle_raw <- log2(counts(dds)/rawc)\n",
    "    rlem_raw <- melt(t(rle_raw))\n",
    "    rlem_raw <- merge(rlem_raw, sample_info, by.x=\"X1\", by.y=0, all=T)\n",
    "\n",
    "    p1 <- ggplot(rlem_raw, aes(y=value, x=X1, colour=condition)) + geom_boxplot() + \n",
    "                    theme(axis.text.x=element_blank()) + \n",
    "                    geom_hline(yintercept=0, lty=\"dashed\", colour=\"black\") +\n",
    "                    labs(title=\"Raw counts\", y=\"\", x=\"\") +\n",
    "                    scale_colour_manual(values=c(Palette)) +\n",
    "                    theme(legend.direction=\"horizontal\")\n",
    "\n",
    "    # norm counts\n",
    "    normc <- apply(counts(dds, normalized=TRUE),1,median)\n",
    "    rle_norm <- log2(counts(dds, normalized=TRUE)/normc)\n",
    "    rlem_norm <- melt(t(rle_norm))\n",
    "    rlem_norm <- merge(rlem_norm, sample_info, by.x=\"X1\", by.y=0, all=T)\n",
    "    head(rlem_norm)\n",
    "\n",
    "    p2 <- ggplot(rlem_norm, aes(y=value, x=X1, colour=condition)) + geom_boxplot() + \n",
    "                    theme(axis.text.x=element_blank()) + \n",
    "                    geom_hline(yintercept=0, lty=\"dashed\", colour=\"black\") +\n",
    "                    labs(title=\"Normalised counts\", y=\"\", x=\"\") +\n",
    "                    scale_colour_manual(values=c(Palette))\n",
    "\n",
    "    # Norm counts with filtering (Cooks distance, independent filtering)\n",
    "    filtc <- apply(filt_norm_counts,1,median)\n",
    "    rle_filt <- log2(filt_norm_counts/filtc)\n",
    "    rlem_filt <- melt(t(rle_filt))\n",
    "    rlem_filt <- merge(rlem_filt, sample_info, by.x=\"X1\", by.y=0, all=T)\n",
    "    # head(rlem_filt)\n",
    "\n",
    "    p3 <- ggplot(rlem_filt, aes(y=value, x=X1, colour=condition)) + geom_boxplot() + \n",
    "                    theme(axis.text.x=element_blank()) + \n",
    "                    geom_hline(yintercept=0, lty=\"dashed\", colour=\"black\") +\n",
    "                    labs(title=\"Filtered normalised counts\", y=\"\", x=\"\") +\n",
    "                    scale_colour_manual(values=c(Palette))\n",
    "\n",
    "    # VST counts\n",
    "    vstc <- apply(vstMat,1,median)\n",
    "    rle_vst <- log2(vstMat/vstc)\n",
    "    rlem_vst <- melt(t(rle_vst))\n",
    "    rlem_vst <- merge(rlem_vst, sample_info, by.x=\"X1\", by.y=0, all=T)\n",
    "\n",
    "    p4 <- ggplot(rlem_vst, aes(y=value, x=X1, colour=condition)) + geom_boxplot() + \n",
    "                    theme(axis.text.x=element_blank()) + \n",
    "                    geom_hline(yintercept=0, lty=\"dashed\", colour=\"black\") +\n",
    "                    labs(title=\"VST counts\", y=\"\", x=\"\") +\n",
    "                    scale_colour_manual(values=c(Palette)) +\n",
    "                    theme(legend.direction=\"horizontal\")\n",
    "\n",
    "    get_legend <- function(a.gplot){ \n",
    "    tmp <- ggplot_gtable(ggplot_build(a.gplot)) \n",
    "    leg <- which(sapply(tmp$grobs, function(x) x$name) == \"guide-box\") \n",
    "    legend <- tmp$grobs[[leg]] \n",
    "    return(legend)}\n",
    "\n",
    "    key <- get_legend(p4)\n",
    "    \n",
    "    plots = list(p1, p2, p3, p4, key)\n",
    "    \n",
    "    return(plots)\n",
    "}\n",
    "\n",
    "plots <- plot_countRLE(dds, sample_info, filt_norm_counts)\n",
    "\n",
    "a <- plots[[1]]   \n",
    "b <- plots[[2]] \n",
    "c <- plots[[3]]   \n",
    "d <- plots[[4]] \n",
    "e <- plots[[5]] \n",
    "\n",
    "grid.arrange(a + theme(legend.position=\"none\"), \n",
    "             b + theme(legend.position=\"none\"), \n",
    "             c + theme(legend.position=\"none\"), \n",
    "             d + theme(legend.position=\"none\"), \n",
    "             ncol=2, nrow=2, bottom=e,\n",
    "             left=textGrob(\"Relative Log Expression\", gp=gpar(fontsize=21, fontfamily=\"Helvetica\"), rot=90))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Exploratory Analysis**\n",
    "### Dimensionality Reduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o sigVst\n",
    "\n",
    "# get significant VST counts\n",
    "sigvar_peaks <- rownames(res)[res$padj<0.05 & !is.na(res$padj)] # get sig DE genes\n",
    "sigVst <- vstMat[sigvar_peaks,]\n",
    "\n",
    "sigVst <- as.data.frame(sigVst)\n",
    "# sigVst\n",
    "# head(sigVst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -w 1200 -h 450\n",
    "                      \n",
    "pca <- prcomp(t(sigVst), scale=FALSE)\n",
    "\n",
    "pca_plots <- ggplot_prcomp(pca, \n",
    "             plots=list(\"A\"=c(\"PC1\",\"PC2\"), \"B\"=c(\"PC3\", \"PC4\")),\n",
    "             sample_information=sample_info, \n",
    "             shape=\"replicate\", \n",
    "             color=\"condition\",\n",
    "#              label=\"replicate\",\n",
    "             size=6,\n",
    "             nudge_scale_factor=30) \n",
    "\n",
    "a <- pca_plots$A\n",
    "b <- pca_plots$B\n",
    "c <- pca_plots$scree\n",
    "\n",
    "a = a + scale_colour_manual(values=Palette) +\n",
    "        guides(colour=guide_legend(override.aes=list(size=8)), shape=guide_legend(override.aes=list(size=8)))\n",
    "               \n",
    "b = b + scale_colour_manual(values=Palette) \n",
    "\n",
    "c = c \n",
    "\n",
    "get_legend <- function(a.gplot){ \n",
    "tmp <- ggplot_gtable(ggplot_build(a.gplot)) \n",
    "leg <- which(sapply(tmp$grobs, function(x) x$name) == \"guide-box\") \n",
    "legend <- tmp$grobs[[leg]] \n",
    "return(legend)}\n",
    "\n",
    "a <- a + theme(legend.direction=\"horizontal\", legend.box=\"horizontal\")\n",
    "legend <- get_legend(a)\n",
    "                      \n",
    "grid.arrange(top=textGrob(\"PCA of VST Counts at Differentially Accessible Peaks\", \n",
    "                          gp=gpar(fontface=\"bold\", fontfamily=\"Helvetica\", fontsize=23)),\n",
    "             a + theme(legend.position=\"none\"), b + theme(legend.position=\"none\"), c, \n",
    "             bottom=legend, ncol=3, nrow=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -w 450 -h 400\n",
    "\n",
    "library(Rtsne)\n",
    "tsne_out = Rtsne(t(sigVst), pca=T, perplexity=2)\n",
    "\n",
    "tsne_df <- as.data.frame(tsne_out$Y)\n",
    "rownames(tsne_df) <- colnames(t(sample_info))\n",
    "colnames(tsne_df) <- c(\"tSNE1\", \"tSNE2\")\n",
    "\n",
    "tsne_df <- merge(tsne_df, sample_info, by=0)\n",
    "\n",
    "tsne_plot <- ggplot(tsne_df, aes(y=tSNE1, x=tSNE2, colour=condition, shape=factor(replicate))) + \n",
    "                    geom_point(size=6) + \n",
    "                    scale_colour_manual(values=Palette) +\n",
    "                    scale_shape_manual(values=c(16,17,22,23,25),name=\"replicate\")\n",
    "\n",
    "grid.arrange(top=textGrob(\"tSNE of VST Counts at \\nDifferentially Accessible Peaks\", \n",
    "             gp=gpar(fontface=\"bold\", fontfamily=\"Helvetica\", fontsize=18)), \n",
    "             tsne_plot, ncol=1, nrow=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Heatmap of all Differentially Accessible Peaks \n",
    "* VST counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get peak to gene annotations\n",
    "peak2gene = DB.fetch_DataFrame('''select distinct a.gene_name, a.gene_id, b.peak_id from \n",
    "                               ensemblGeneset a, merged_peaks_GREAT_annotated b \n",
    "                               where a.gene_id = b.gene_id''', db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%R -i sigVst,peak2gene -w 600 -h 800\n",
    "\n",
    "mat <- merge(sigVst, peak2gene[c(\"peak_id\", \"gene_name\")], by.x=0, by.y=\"peak_id\")\n",
    "mat <- mat[, !names(mat) %in% c(\"Row.names\")]\n",
    "\n",
    "plot_heatmap(mat, clust_method=\"ward.D2\", dist_method=\"manhattan\", low_opt=0.1, \n",
    "             med_opt=0.5, high_opt=0.9, labels=\"random\", no_labels=50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DE Peaks - Contrasts\n",
    "* Pairwise comparisons of differentially accessible peaks\n",
    "* Adjusted p-value < 0.05 & fold change > 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o condition_pairs\n",
    "\n",
    "# get pairs of sample comparisons\n",
    "a <- unique(sample_info$condition)\n",
    "b <- combn(a, 2, simplify=FALSE)\n",
    "condition_pairs <- as.data.frame(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reformat df\n",
    "condition_pairs = condition_pairs.transpose()\n",
    "condition_pairs.reset_index(drop=True, inplace=True)\n",
    "condition_pairs.columns = [\"a\", \"b\"]\n",
    "condition_pairs[\"comparison\"] = condition_pairs[\"a\"] + \"_VS_\" + condition_pairs[\"b\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -h 550 -w 1200 -i condition_pairs\n",
    "\n",
    "MA <- function(df, title=\"\", gene_info=gene_info){\n",
    "      \n",
    "    df <- merge(df, gene_info[c(\"gene_id\", \"gene_name\", \"peak_id\")], by.x=0, by.y=\"peak_id\")\n",
    "    \n",
    "    MA <- ggplot(subset(df, padj > 0.05 | is.na(padj)), aes(y=log2FoldChange, x=log2(baseMean))) + \n",
    "            geom_point(alpha=0.4, colour=\"gray20\") +\n",
    "            geom_point(data=subset(df, padj < 0.05 & log2FoldChange < 1 | padj < 0.05 & log2FoldChange > -1), \n",
    "                       aes(y=log2FoldChange, x=log2(baseMean)), alpha=0.4, colour=\"gray20\") +\n",
    "            geom_point(data=subset(df, padj < 0.05 & log2FoldChange > 1 | padj < 0.05 & log2FoldChange < -1), \n",
    "                       aes(y=log2FoldChange, x=log2(baseMean)), alpha=0.4, colour=\"red3\") +\n",
    "            scale_y_continuous(limits=c(min(df$log2FoldChange)-2, max(df$log2FoldChange)+2)) +\n",
    "            scale_x_continuous(limits=c(0, max(log2(df$baseMean)))) +\n",
    "            labs(y=\"log2(FoldChange)\", title=title) +\n",
    "            geom_hline(yintercept=0, lty=\"dashed\", col=\"black\") +\n",
    "            geom_label_repel(data=subset(df[df$padj < quantile(df$padj, 0.0005, na.rm=TRUE), ], padj < 0.05),\n",
    "                             aes(y=log2FoldChange, x=log2(baseMean), \n",
    "                             label=gene_name), size=5, \n",
    "                             label.padding=unit(0.3, \"lines\"), label.r=unit(0.5, \"lines\"), segment.color = \"gray20\", \n",
    "                             fill=\"gray20\", colour=\"white\", fontface=\"bold\", label.size=NA, alpha=0.8)\n",
    "\n",
    "    return(MA)\n",
    "}\n",
    "\n",
    "volcano <- function(df, title=\"\", gene_info=gene_info){\n",
    "      \n",
    "    df <- merge(df, gene_info[c(\"gene_id\", \"gene_name\", \"peak_id\")], by.x=0, by.y=\"peak_id\")\n",
    "    \n",
    "    volcano <- ggplot(subset(df, padj > 0.05 | is.na(padj)), aes(y=log2(padj), x=log2FoldChange)) + \n",
    "            geom_point(alpha=0.4, colour=\"gray20\") +\n",
    "            geom_point(data=subset(df, padj < 0.05 & log2FoldChange < 1 | padj < 0.05 & log2FoldChange > -1), \n",
    "                       aes(y=log2(padj), x=log2FoldChange), alpha=0.4, colour=\"gray20\") +\n",
    "            geom_point(data=subset(df, padj < 0.05 & log2FoldChange > 1 | padj < 0.05 & log2FoldChange < -1), \n",
    "                       aes(y=log2(padj), x=log2FoldChange), alpha=0.4, colour=\"red3\") +\n",
    "            labs(y=\"log2(FDR)\", title=title) +\n",
    "            scale_y_reverse(limits=c(0, min(log2(df$padj)))) +\n",
    "            scale_x_continuous(limits=c(min(df$log2FoldChange), max(df$log2FoldChange))) +\n",
    "            geom_vline(xintercept=0, lty=\"dashed\", col=\"black\") +\n",
    "            geom_label_repel(data=subset(df[df$padj < quantile(df$padj, 0.0005, na.rm=TRUE), ], padj < 0.05),\n",
    "                             aes(y=log2(padj), x=log2FoldChange, \n",
    "                             label=gene_name), size=5, \n",
    "                             label.padding=unit(0.3, \"lines\"), label.r=unit(0.5, \"lines\"), segment.color = \"gray20\", \n",
    "                             fill=\"gray20\", colour=\"white\", fontface=\"bold\", label.size=NA, alpha=0.8)\n",
    "\n",
    "    return(volcano)\n",
    "}\n",
    "\n",
    "n = 0\n",
    "for (line in 1:nrow(condition_pairs)){\n",
    "    n = n + 1\n",
    "\n",
    "    sample = condition_pairs[n, ]\n",
    "\n",
    "    a <- as.character(sample$a)\n",
    "    b <- as.character(sample$b)\n",
    "    title <- as.character(sample$comparison)\n",
    "\n",
    "    plot <- as.data.frame(results(dds, contrast=c(\"condition\", a, b)))\n",
    "    p1 <- (MA(plot, title, gene_info=peak2gene))\n",
    "    p2 <- (volcano(plot, title, gene_info=peak2gene))\n",
    "\n",
    "    grid.arrange(p1, p2, ncol=2, nrow=1)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary Table \n",
    "* Peaks with adjusted p-value < 0.05 and fold change > 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o sig_peaks\n",
    "\n",
    "# get no. sig diff. accessible peaks\n",
    "\n",
    "n = 0\n",
    "contrasts = list()\n",
    "for (line in 1:nrow(condition_pairs)){\n",
    "    n = n + 1\n",
    "\n",
    "    sample = condition_pairs[n, ]\n",
    "\n",
    "    a <- as.character(sample$a)\n",
    "    b <- as.character(sample$b)\n",
    "    title <- as.character(sample$comparison)\n",
    "    \n",
    "    if (length(title)[1] > 0){\n",
    "        contrasts[[title]] <- c(a, b)\n",
    "    }\n",
    "\n",
    "    # get significant results\n",
    "    df <- results(dds, contrast=c(\"condition\", a, b))\n",
    "    res <- subset(df, padj < 0.05 & log2FoldChange > 1 |  padj < 0.05 & log2FoldChange < -1)\n",
    "    upreg <- subset(res, log2FoldChange > 0)\n",
    "    downreg <- subset(res, log2FoldChange < 0)\n",
    "    \n",
    "    if (n == 1) {\n",
    "        sig_peaks <- as.data.frame(c(length(res$padj), length(upreg$padj), length(downreg$padj), title), byrow=F)\n",
    "        sig_peaks <- t(sig_peaks)\n",
    "    } else{\n",
    "        df <- as.data.frame(c(length(res$padj), length(upreg$padj), length(downreg$padj), title), byrow=F)\n",
    "        df <- t(df)\n",
    "        sig_peaks <- rbind(sig_peaks, df)\n",
    "    }\n",
    "\n",
    "}\n",
    "\n",
    "sig_peaks <- as.data.frame(sig_peaks)\n",
    "rownames(sig_peaks) <- NULL\n",
    "colnames(sig_peaks) <- c(\"No. Peaks\", \"Upreg\", \"Downreg\", \"Contrast\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_peaks[\"No. Peaks\"] = sig_peaks[\"No. Peaks\"].apply(lambda x: int(x))\n",
    "sig_peaks.sort_values(\"No. Peaks\", ascending=False).head(len(sig_peaks))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DESeq2 results are available at:\n",
    "* Differentially accessible peaks saved in \"DESeq2_results\" table in csvdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o acc_df\n",
    "\n",
    "# make table of all significant results for each comparison & upload to db\n",
    "\n",
    "n = 0\n",
    "contrasts = list()\n",
    "for (line in 1:nrow(condition_pairs)){\n",
    "    n = n + 1\n",
    "\n",
    "    sample = condition_pairs[n, ]\n",
    "\n",
    "    a <- as.character(sample$a)\n",
    "    b <- as.character(sample$b)\n",
    "    title <- as.character(sample$comparison)\n",
    "    \n",
    "    if (length(title)[1] > 0){\n",
    "        contrasts[[title]] <- c(a, b)\n",
    "    }\n",
    "\n",
    "    # get significant results\n",
    "    res <- results(dds, contrast=c(\"condition\", a, b))\n",
    "    sig_res <- subset(res, padj < 0.05 & log2FoldChange > 1 |  padj < 0.05 & log2FoldChange < -1)\n",
    "    upreg <- as.data.frame(subset(sig_res, log2FoldChange > 0))\n",
    "    downreg <- as.data.frame(subset(sig_res, log2FoldChange < 0))\n",
    "\n",
    "    if (length(t(upreg)) > 0){\n",
    "        upreg$accessibility <- \"increase\"\n",
    "        upreg$comparison <- title\n",
    "        upreg$peak_id <- rownames(upreg)\n",
    "    } #else{\n",
    "#         print(paste(\"No. upreg peaks with padj < 0.05 & fold change > 2 in comparison:\", title))\n",
    "#     }\n",
    "    \n",
    "    if (length(t(downreg)) > 0){\n",
    "        downreg$accessibility <- \"decrease\"\n",
    "        downreg$comparison <- title\n",
    "        downreg$peak_id <- rownames(downreg)\n",
    "    } #else{\n",
    "#         print(paste(\"No. downreg peaks with padj < 0.05 & fold change > 2 in comparison:\", title))\n",
    "#     }\n",
    "    \n",
    "    if (n == 1) {\n",
    "            acc_df <- rbind(upreg, downreg)\n",
    "    } else{\n",
    "        df <- rbind(upreg, downreg)\n",
    "        acc_df <- rbind(acc_df, df)\n",
    "    }\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def upload2csvdb(df, peak2gene, db):\n",
    "    '''Upload significant DESeq2 results to csvdb'''\n",
    "    \n",
    "    df = pd.merge(df, peak2gene[[\"gene_name\", \"peak_id\"]].drop_duplicates(), \n",
    "             how=\"inner\", on=\"peak_id\")\n",
    "    \n",
    "    connect = sqlite3.connect(db)\n",
    "    \n",
    "    df.to_sql(\"DESeq2_results\", connect, if_exists=\"replace\", index=False)\n",
    "    \n",
    "upload2csvdb(acc_df, peak2gene, db)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* BED files of differentially accessible peaks from each comparison located in \"DESeq2_peaks.dir/\"\n",
    "    * subdivded into increasing & decreasing accessibility\n",
    "    * columns = contig, peak_start, peak_end, peak_id, -log10(padj), peak_summit, gene_name, accessibility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "statement = '''select a.contig, a.peak_start, a.peak_end, a.peak_id, \n",
    "                b.gene_name, b.comparison, b.accessibility, b.padj \n",
    "                from merged_peaks_GREAT_annotated a, \n",
    "                DESeq2_results b where a.peak_id=b.peak_id '''\n",
    "\n",
    "dfs = DB.fetch_DataFrame(statement, db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "if [ ! -d DESeq2_peaks.dir ] ; then mkdir DESeq2_peaks.dir; fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for contrast in sig_peaks[\"Contrast\"]:\n",
    "    for value in [\"increase\", \"decrease\"]:\n",
    "        \n",
    "        statement = '''select a.contig, a.peak_start, a.peak_end, a.peak_id, \n",
    "                    b.padj, a.peak_centre, b.gene_name, b.accessibility\n",
    "                    from merged_peaks_GREAT_annotated a, \n",
    "                    DESeq2_results b where a.peak_id=b.peak_id \n",
    "                    and b.comparison=\"%(contrast)s\" and b.accessibility=\"%(value)s\"''' % locals()\n",
    "\n",
    "        df = DB.fetch_DataFrame(statement, db)\n",
    "\n",
    "        if len(df) > 0:\n",
    "            \n",
    "            # convert padj to -log10 scale so that it has a comparable score to macs2 qvalue\n",
    "            # e.g. can sort on significance high to low, rather than low to high\n",
    "            df[\"padj\"] = df[\"padj\"].apply(lambda x: -np.log10(x))\n",
    "\n",
    "            # name for bed file\n",
    "            outdir = \"DESeq2_peaks.dir/\"\n",
    "            reg = ''.join(pd.unique(df[\"accessibility\"]))\n",
    "            outfile = outdir + '_'.join([\"DE_peaks\", contrast, reg]) + \".bed\" \n",
    "           \n",
    "            df.to_csv(outfile, sep=\"\\t\", header=False, index=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
